metron-platform/metron-writer/src/main/java/org/apache/metron/writer/BulkWriterComponent.java

```diff
@@ -118,12 +118,15 @@ public void commit(BulkWriterResponse response) {
 
   public void error(String sensorType, Throwable e, Iterable<Tuple> tuples, MessageGetStrategy messageGetStrategy) {
     LOG.error(format("Failing %d tuple(s); sensorType=%s", Iterables.size(tuples), sensorType), e);
-    MetronError error = new MetronError()
-            .withSensorType(Collections.singleton(sensorType))
-            .withErrorType(Constants.ErrorType.INDEXING_ERROR)
-            .withThrowable(e);
-    tuples.forEach(t -> error.addRawMessage(messageGetStrategy.get(t)));
-    handleError(tuples, error);
+    tuples.forEach(t -> {
```

Comment:
Seems like this will ack the same tuples repetitively.  If 500 messages in a batch fail, then we will ack all 500 of them, 500 times.

There is also the nuisance that we report the error to the collector for each and every failed message, instead of just once for the batch.  There is only one `Throwable` error to report, so we should just report it once.

We may need something like this.

```suggestion
    // emit one error for each failed message
    tuples.forEach(t -> {
      MetronError error = new MetronError()
              .withSensorType(Collections.singleton(sensorType))
              .withErrorType(Constants.ErrorType.INDEXING_ERROR)
              .withThrowable(e)
              .addRawMessage(messageGetStrategy.get(t));
      collector.emit(Constants.ERROR_STREAM, new Values(error.getJSONObject()));
      collector.ack(t);
    });

    // there is only one error to report for all of the failed tuples
    collector.reportError(e);
  }
```

Reply:
You're right!  My mistake.  I changed it to pass in just the current tuple to `handleError`.  Let me know if that works.

Reply:
It might also be useful to discuss on this PR, if the other similar method `error(Throwable, Iterable<Tuple>)`  needs updated.  That method gets called if messages are unparsable.  And in that case we send only a single error, instead of one for each error'd message.  Is that what we want?

It might be what we want, but at the very least it would be useful to comment in there as to why we treat it differently, if indeed it should be different.  Could be something like this.
```
  public void error(Throwable e, Iterable<Tuple> tuples) {
    LOG.error(format("Failing %d tuple(s) due to invalid message; sensorType unknown", Iterables.size(tuples)), e);

    // emit only a single error message, even though many failed because ...
    MetronError error = new MetronError()
            .withErrorType(Constants.ErrorType.INDEXING_ERROR)
            .withThrowable(e);
    collector.emit(Constants.ERROR_STREAM, new Values(error.getJSONObject()));
    tuples.forEach(t -> collector.ack(t));

    // there is only one error to report for all of the failed tuples
    collector.reportError(e);
  }
```

Reply:
> You're right! My mistake. I changed it to pass in just the current tuple to handleError

Your latest commit was my first thought of a fix too; it is much simpler.  But I don't think it solves the second problem that I mentioned.  

> There is also the nuisance that we report the error to the collector for each and every failed message, instead of just once for the batch. There is only one Throwable error to report, so we should just report it once.

I think reporting the error once is both a usability and performance improvement.

I had to unravel the code base a bit more in my suggested code snippet so that it only reports the error once.

Reply:
You're right again.  I agree, we shouldn't be reporting the same exception for every message in the batch.  I changed it to what you suggested in the initial comment and updated the tests.

For the other issue you bring up with `error(Throwable, Iterable<Tuple>)`, it looks like it gets called from only one place:  `BulkMessageWriterBolt.handleMissingMessage`.  This method only deals with a single tuple so it is misleading that the `error` method accepts multiple tuples.  Would changing that to accept a single tuple make it less confusing?

Reply:
> This method only deals with a single tuple so it is misleading that the error method accepts multiple tuples. Would changing that to accept a single tuple make it less confusing?

Yes, agreed.  That would help clarify things.

Reply:
Done.

Reply:
Looks great.  I'm just going to spin it up and do some manual testing now.